Node: Add type inference for TransformResult

ArnaudBarre opened the pull request:
This PR improves nullability of transform and bundle outputs.

```ts
let {code, map} = css.transform({
  filename: 'style.css',
  code: Buffer.from('.foo { color: red }'),
  sourceMap: true,
});
console.log(map.toString()); // Valid in TS without assertion 🎉
```

Note: Prettier seems to not be applied on this file, so I tried to follow formatting conventions manually.

devongovett commented:
Could you explain a little more about what this does? I've used TS but this is next level stuff. 😅

ArnaudBarre commented:
I changed to use function overloads. The first solution was more scalable & able to narrow down for false cases, but TS became too lax on parameters and not warnings on bad parameter.

I added tests via tsd (now recommended by MS) that will explain the goal.

Tell me if you want me to do others methods & add more tests or remove the setup for tsd (for information, ~50M node_modules, ~3s run on Mac without m1)

ArnaudBarre commented:
I quickly tested the script and the generics syntax seems to create syntax errors.
Using this [tool](https://transform.tools/typescript-to-flow), we would need to change this two lines into:
```
declare type WithFlag<Options, Key: $Keys<Options>> = {
  ...Options,
  ...$ObjMapi<{ [k: Key]: any }, <key>(key) => true>,
};
declare type NonNullableTransformResult<Key: $Keys<TransformResult>> = {
  ...TransformResult,
  ...$ObjMapi<
    { [k: Key]: any },
    <key>(key) => Exclude<$ElementType<TransformResult, key>, void>
    >,
};
```
But I can't say if it's valid, I don't know flow at all.
